module/localisation/RobotLocalisation/src/RobotLocalisation.cpp

```diff
@@ -0,0 +1,171 @@
+            cfg.ukf.initial_covariance.velocity = config["ukf"]["initial_covariance"]["velocity"].as<Expression>();
+            cfg.association_distance            = config["association_distance"].as<double>();
+            cfg.max_missed_count                = config["max_missed_count"].as<int>();
+        });
+
+
+        on<Trigger<VisionRobots>, With<GreenHorizon>, Single>().then([this](const VisionRobots& vision_robots,
+                                                                            const GreenHorizon& horizon) {
+            // Print tracked_robots ids
+            log<NUClear::DEBUG>("Robots tracked:");
+            for (const auto& tracked_robot : tracked_robots) {
+                log<NUClear::DEBUG>("\tID: ", tracked_robot.id);
+            }
+
+            // Set all tracked robots to unseen
+            for (auto& tracked_robot : tracked_robots) {
+                tracked_robot.seen = false;
+            }
+
+            // **************** Data association ****************
+            if (!vision_robots.robots.empty()) {
+                Eigen::Isometry3d Hwc = Eigen::Isometry3d(vision_robots.Hcw).inverse();
+                for (const auto& vision_robot : vision_robots.robots) {
+                    // Position of robot {r} in world {w} space
+                    auto rRWw = Hwc * vision_robot.rRCc;
+
+                    // Only consider vision measurements within the green horizon
+                    if (point_in_convex_hull(horizon.horizon, rRWw)) {
+                        // Data association: find tracked robot which is associated with the vision measurement
+                        data_association(rRWw);
+                    }
+                }
+            }
+
+            // **************** Robot tracking maintenance ****************
+            for (auto& tracked_robot : tracked_robots) {
+                auto state = RobotModel<double>::StateVec(tracked_robot.ukf.get_state());
+
+                // If a tracked robot has moved outside of view, keep it as seen so we don't lose it
+                // A robot is outside of view if it is not within the green horizon
+                // TODO: It may be better to use fov and image size to determine if a robot should be seen
+                if (!point_in_convex_hull(horizon.horizon, Eigen::Vector3d(state.rRWw.x(), state.rRWw.y(), 0))) {
+                    tracked_robot.seen = true;
+                }
+
+                // If the tracked robot has not been seen, increment the consecutively missed count
+                tracked_robot.missed_count = tracked_robot.seen ? 0 : tracked_robot.missed_count + 1;
+            }
+
+            // Make a vector to store kept robots
+            std::vector<TrackedRobot> new_tracked_robots;
+            // Only keep robots that are not missing or too close to others
+            for (const auto& tracked_robot : tracked_robots) {
```

Comment:
Could consider replacing this with a ranges::filter. An erase remove idiom would be most efficient actually, since you want it on the same vector after the fact.

Reply:
I agree. I believe this requires a bit of effort though in setting up TrackedRobot struct to work nicely with this, I could work on it but not super critical imo